rewrite-java/src/main/java/org/openrewrite/java/cleanup/XmlParserXXEVulnerabilityFix.java

```diff
@@ -0,0 +1,138 @@
+import org.openrewrite.java.JavaParser;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.HasTypes;
+import org.openrewrite.java.tree.*;
+
+import java.util.Collections;
+
+public class XmlParserXXEVulnerabilityFix extends Recipe {
+    private static final MethodMatcher XML_PARSER_FACTORY_INSTANCE_METHOD_MATCHER = new MethodMatcher("javax.xml.stream.XMLInputFactory new*()");
+    private static final MethodMatcher XML_PARSER_FACTORY_SET_PROPERTY_METHOD_MATCHER = new MethodMatcher("javax.xml.stream.XMLInputFactory setProperty(java.lang.String, ..)");
+
+    private static final String XML_FACTORY_FQN = "javax.xml.stream.XMLInputFactory";
+    private static final String SUPPORTING_EXTERNAL_ENTITIES_PROPERTY_NAME = "IS_SUPPORTING_EXTERNAL_ENTITIES";
+    private static final String SUPPORT_DTD_PROPERTY_NAME = "SUPPORT_DTD";
+    private static final String XML_PARSER_INITIALIZATION_METHOD = "xml-parser-initialization-method";
+    private static final String XML_FACTORY_VARIABLE_NAME = "xml-factory-variable-name";
+
+    private static final ThreadLocal<JavaParser> JAVA_PARSER_THREAD_LOCAL = ThreadLocal.withInitial(() -> JavaParser.fromJavaVersion().build());
+
+    @Override
+    public String getDisplayName() {
+        return "XML Parser XXE Vulnerability Fix.";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Avoid exposing dangerous features of the XML parser by setting XMLInputFactory IS_SUPPORTING_EXTERNAL_ENTITIES and SUPPORT_DTD properties to false.";
+    }
+
+    @Override
+    protected @Nullable TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new HasTypes(Collections.singletonList(XML_FACTORY_FQN)).getVisitor();
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext executionContext) {
+                J.ClassDeclaration cd = super.visitClassDeclaration(classDecl, executionContext);
+                Cursor supportsExternalCursor = getCursor().getMessage(SUPPORTING_EXTERNAL_ENTITIES_PROPERTY_NAME);
+                Cursor supportsDTDCursor = getCursor().getMessage(SUPPORT_DTD_PROPERTY_NAME);
+                Cursor initializationCursor = getCursor().getMessage(XML_PARSER_INITIALIZATION_METHOD);
+                String xmlFactoryVariableName = getCursor().getMessage(XML_FACTORY_VARIABLE_NAME);
+
+                Cursor setPropertyBlockCursor = null;
+                if (supportsExternalCursor == null && supportsDTDCursor == null) {
+                    setPropertyBlockCursor = initializationCursor;
+                } else if (supportsExternalCursor == null ^ supportsDTDCursor == null) {
+                    setPropertyBlockCursor = supportsExternalCursor == null ? supportsDTDCursor : supportsExternalCursor;
+                }
```

Comment:
You know, this may be the first time I've ever seen Java's XOR in actual code. It's like seeing a unicorn.

If it's not important to leave `setPropertyBlockCursor` null when _both_ supportsExternalCursor and supportsDTDCursor are non-null, then I think this would be more legible:
```suggestion
                } else if (supportsExternalCursor != null) {
                    setPropertyBlockCursor = supportsExternalCursor;
                } else {
                    setPropertyBlockCursor = supportsDTDCursor;
                }
```

Reply:
I completely agree. It's rare to use the XOR operator,  I have tried several different options for this test, and the XOR seems to be the most simple.

If **both** `supportsExternalCursor` and `supportsDTDCursor` are **non-null**, then the `XmlParser` is not vulnerable and no action is required.  

Because the `setPropertyBlockCursor` is used as a switch for executing the `XmlFactoryInsertPropertyStatementVisitor` it must be **null** when both the `supportsExternalCursor` and `supportsDTDCursor` are **non-null**.